packages/itwinui-css/src/panels/panels.scss

```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+// See LICENSE.md in the project root for license terms and full copyright notice.
+
+.iui-panel-wrapper {
+  position: relative;
+  overflow: hidden;
+}
+
+.iui-panel {
+  position: absolute;
```

Comment:
To make them stack horizontally instead of vertically, you can use regular flex/grid layout. The element would basically scroll horizontally during the animation (1. click, 2. add new inert panel to DOM, 3. scroll to new panel and un-inert it, 4. remove old panel from DOM).

Reply:
Did you try using the scrolling approach I described in my previous comment? If so, did you run into any problems with it?

The general process would go something like:
0. Panel 1 is the only thing in DOM
1. When panel 2 is triggered, add it to DOM but keep it outside the scrollport
3. Use [`scrollIntoView`](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) to move panel 2 into scrollport and panel 1 out of scrollport
4. Remove panel 1 from DOM

I made a diagram to illustrate how this process. I'm using the word "animation", but really it's just smooth scrolling.

![](https://github.com/user-attachments/assets/e31c8a6b-0ce4-4100-b2b5-aeae2dc6b11d)

Reply:
I was able to make it work with the help of scroll snapping. Here are the changes I made against the [current code](https://github.com/iTwin/iTwinUI/tree/39054001a7eb5bd0b6123d9d00e40b5b70b9f23f/packages):

<details><summary>Diff</summary>

```diff
diff --git a/packages/itwinui-css/src/panels/panels.scss b/packages/itwinui-css/src/panels/panels.scss
index f30e35683..d6f5ca122 100644
--- a/packages/itwinui-css/src/panels/panels.scss
+++ b/packages/itwinui-css/src/panels/panels.scss
@@ -2,12 +2,8 @@
 // See LICENSE.md in the project root for license terms and full copyright notice.
 
 .iui-panel-wrapper {
-  overflow: auto;
-  overflow-anchor: auto;
-
-  @media (prefers-reduced-motion: no-preference) {
-    scroll-behavior: smooth;
-  }
+  overflow: hidden;
+  scroll-snap-type: x proximity;
 }
 
 // Intentional selector to increase specificity
@@ -19,4 +15,8 @@
 .iui-panel {
   inline-size: 100%;
   flex-shrink: 0;
+
+  &:not([data-iui-transitioning='true']) {
+    scroll-snap-align: end;
+  }
 }
diff --git a/packages/itwinui-react/src/core/Panels/Panels.tsx b/packages/itwinui-react/src/core/Panels/Panels.tsx
index 042636d51..63a9255fe 100644
--- a/packages/itwinui-react/src/core/Panels/Panels.tsx
+++ b/packages/itwinui-react/src/core/Panels/Panels.tsx
@@ -23,7 +23,6 @@ import cx from 'classnames';
 import {
   PanelsInstanceContext,
   PanelsInstanceProvider,
-  panelsReducer,
   PanelsWrapperContext,
 } from './helpers.js';
 import type { PanelsInstance, TriggerMapEntry } from './helpers.js';
@@ -107,18 +106,7 @@ const PanelsWrapper = React.forwardRef((props, forwardedRef) => {
 
   const ref = React.useRef<HTMLDivElement | null>(null);
 
-  const [{ activePanel, nextPanel }, dispatch] = React.useReducer(
-    panelsReducer,
-    {
-      activePanel: initialActiveId,
-      nextPanel: undefined,
-      onActiveIdChange,
-    },
-  );
-
-  React.useEffect(() => {
-    console.log('panels', activePanel, nextPanel);
-  }, [activePanel, nextPanel]);
+  const [activePanel, setActivePanel] = React.useState(initialActiveId);
 
   const [_panelElements, setPanelElements] = React.useState<
     Record<string, HTMLElement | null>
@@ -129,33 +117,6 @@ const PanelsWrapper = React.forwardRef((props, forwardedRef) => {
     Record<string, TriggerMapEntry>
   >({});
 
-  React.useEffect(() => {
-    const refCurrent = ref.current;
-
-    const scrollEndListenerCurrent = () => {
-      dispatch({ type: 'end-panel-transition' });
-    };
-
-    const scrollListener = () => {
-      clearTimeout((window as any).scrollEndTimer);
-      (window as any).scrollEndTimer = setTimeout(
-        scrollEndListenerCurrent,
-        100,
-      );
-    };
-
-    if ('onscrollend' in window) {
-      refCurrent?.addEventListener('scrollend', scrollEndListenerCurrent);
-    } else {
-      refCurrent?.addEventListener('scroll', scrollListener);
-    }
-
-    return () => {
-      refCurrent?.removeEventListener('scrollend', scrollEndListenerCurrent);
-      refCurrent?.removeEventListener('scroll', scrollListener);
-    };
-  }, [activePanel, nextPanel]);
-
   const changeActivePanel = React.useCallback(
     async (newActiveId: string) => {
       if (
@@ -167,29 +128,15 @@ const PanelsWrapper = React.forwardRef((props, forwardedRef) => {
         return;
       }
 
-      ReactDOM.flushSync(() => {
-        // setNextPanel(newActiveId);
-        dispatch({ type: 'start-panel-transition', nextPanel: newActiveId });
+      ReactDOM.flushSync(() => setActivePanel(newActiveId));
+      panelElements.current?.[newActiveId]?.scrollIntoView({
+        block: 'nearest',
+        inline: 'center',
+        behavior: 'smooth',
       });
 
-      // setTimeout(() => {
-      //   console.log('HERE', panelElements.current[activePanel]);
-      //   panelElements.current[activePanel]?.scrollIntoView({
-      //     behavior: 'instant',
-      //     block: 'nearest',
-      //     inline: 'center',
-      //   });
-      // }, 1_000);
-
       await new Promise((resolve) => {
         setTimeout(() => {
-          console.log('Scrolling', panelElements.current?.[newActiveId]);
-
-          panelElements.current?.[newActiveId]?.scrollIntoView({
-            block: 'nearest',
-            inline: 'center',
-          });
-
           // WIP Focus
           // const activePanelElement = document.getElementById(activePanel);
           // const newActiveElement = document.getElementById(newActiveId);
@@ -214,9 +161,8 @@ const PanelsWrapper = React.forwardRef((props, forwardedRef) => {
           //     trigger?.focus({ preventScroll: true });
           //   }
           // }
-
           resolve(null);
-        }, 4_000);
+        });
       });
 
       // TODO: focus
@@ -247,14 +193,13 @@ const PanelsWrapper = React.forwardRef((props, forwardedRef) => {
       value={React.useMemo(
         () => ({
           activePanel,
-          nextPanel,
           changeActivePanel,
           triggers,
           setTriggers,
           panelElements: panelElements.current,
           setPanelElements,
         }),
-        [activePanel, changeActivePanel, panelElements, nextPanel, triggers],
+        [activePanel, changeActivePanel, panelElements, triggers],
       )}
     >
       <PanelsInstanceProvider instance={instance}>
@@ -295,7 +240,7 @@ type PanelProps = {
 const Panel = React.forwardRef((props, forwardedRef) => {
   const { id: idProp, children, className, ...rest } = props;
 
-  const { activePanel, nextPanel, triggers, panelElements, setPanelElements } =
+  const { activePanel, triggers, panelElements, setPanelElements } =
     React.useContext(PanelsWrapperContext) || {};
 
   const fallbackId = React.useId();
@@ -328,9 +273,14 @@ const Panel = React.forwardRef((props, forwardedRef) => {
     [id, triggers],
   );
 
-  const isMounted = [activePanel, nextPanel].includes(id);
+  const previousActivePanel = useDelayed(activePanel);
+
+  const isMounted = [activePanel, previousActivePanel].includes(id);
+  const isTransitioning =
+    activePanel === id && activePanel !== previousActivePanel;
+
   // TODO: Being back inert checks
-  // const isInert = [nextPanel].includes(id);
+  const isInert = previousActivePanel === id && activePanel !== id;
 
   const refs = useMergedRefs(setElement, forwardedRef);
 
@@ -348,7 +298,8 @@ const Panel = React.forwardRef((props, forwardedRef) => {
           className={cx('iui-panel', className)}
           aria-labelledby={`${id}-header`}
           tabIndex={-1}
-          // {...{ inert: isInert ? '' : undefined }}
+          {...{ inert: isInert ? 'true' : undefined }}
+          data-iui-transitioning={isTransitioning ? 'true' : undefined}
           {...rest}
         >
           {children}
@@ -520,3 +471,22 @@ export const Panels = {
   BackButton: PanelBackButton,
   useInstance: useInstance as () => PanelsInstance,
 };
+
+/**
+ * Returns a copy of value which reflects state changes after a set delay.
+ */
+function useDelayed<T>(value: T, { delay } = { delay: 500 }): T | undefined {
+  const [delayed, setDelayed] = React.useState<T | undefined>(undefined);
+  const timeout = React.useRef<ReturnType<typeof setTimeout> | undefined>(
+    undefined,
+  );
+
+  React.useEffect(() => {
+    timeout.current = setTimeout(() => setDelayed(value), delay);
+    return () => {
+      clearTimeout(timeout.current);
+    };
+  }, [value, delay]);
+
+  return delayed;
+}
diff --git a/packages/itwinui-react/src/core/Panels/helpers.tsx b/packages/itwinui-react/src/core/Panels/helpers.tsx
index 1107cf971..f85d9f889 100644
--- a/packages/itwinui-react/src/core/Panels/helpers.tsx
+++ b/packages/itwinui-react/src/core/Panels/helpers.tsx
@@ -4,7 +4,6 @@
  *--------------------------------------------------------------------------------------------*/
 import * as React from 'react';
 import { useSynchronizeInstance } from '../../utils/index.js';
-import type { PanelsWrapperProps } from './Panels.js';
 
 export type PanelsInstance = {
   /** Go back to the panel that has a trigger that points to the current panel. */
@@ -19,7 +18,6 @@ export const PanelsWrapperContext = React.createContext<
   | {
       // TODO: Change names to like activePanelId. But maybe we can use elements directly without ids?
       activePanel: string;
-      nextPanel?: string;
       triggers: Record<string, TriggerMapEntry>;
       setTriggers: React.Dispatch<
         React.SetStateAction<Record<string, TriggerMapEntry>>
@@ -83,47 +81,3 @@ export const PanelsInstanceProvider = (props: PanelInstanceProviderProps) => {
     </PanelsInstanceContext.Provider>
   );
 };
-
-// ----------------------------------------------------------------------------
-
-type PanelsReducerState = {
-  activePanel: string;
-  nextPanel?: string;
-  onActiveIdChange?: PanelsWrapperProps['onActiveIdChange'];
-};
-
-type PanelsReducerAction =
-  | {
-      type: 'start-panel-transition';
-      nextPanel: string;
-    }
-  | {
-      type: 'end-panel-transition';
-    };
-
-export const panelsReducer = (
-  state: PanelsReducerState,
-  action: PanelsReducerAction,
-) => {
-  if (action.type === 'start-panel-transition') {
-    return {
-      ...state,
-      nextPanel: action.nextPanel,
-    };
-  }
-
-  if (action.type === 'end-panel-transition') {
-    if (state.nextPanel == null) {
-      return state;
-    }
-
-    const newActivePanel = state.nextPanel;
-
-    state.onActiveIdChange?.(newActivePanel);
-    return {
-      activePanel: newActivePanel,
-    };
-  }
-
-  return state;
-};

```


</details>

Reply:
> Is the 500ms arbitrary? I added a check in [`e0c8590`](https://github.com/iTwin/iTwinUI/pull/2001/commits/e0c8590cfd148e938165ac1a4753797000565b9d) to prevent a page change from happening even if the scroll has finished before 500ms.

Yes it is arbitrary, because the exact number doesn't matter. I'm not sure why you want to prevent a panel change; is there a problem with letting the user quickly switch back and forth between panels?

Reply:
Yes, I think it is fine to not animate for rapid clicks. This is exactly what happens in `Carousel` too. The important part is to always respect the user's clicks. This is why `activePanel` should be the only source of truth.